### PR TITLE
RD-970 Reports

### PR DESCRIPTION
**Pull request by aj-hitchins**
# Reports #

## Description ##

Added reporting for gap line detection.

Jira ticket : [RD-970](https://hummingbirdtech.atlassian.net/browse/RD-970)

## How To Test ##

`hb-gap-detection a55dac69-01ed-42cb-be87-794bf953b919`

The report will be available in your tmp folder

## Checklist ##

- [ x] PEP8 compliant (run `black --check $(git diff --name-only HEAD origin/develop | grep .py) -l 119`);
- [x ] Passing tests;
- [ ] _(If required)_ Confluence pages updated.

### CODE REVIEW

File: hummingbirdtech/planting_lines/gap_detection/report.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        :type lines: List[shapely.geometry.MultiLineString]
+        :rtype: float
+        """
+
+    def _get_samples(self, rgb_image, row_band, vegetation_band, crop_row_lines, gap_lines):
+        """
+        :type rgb_image: hummingbirdtech.domain.image.Image
+        :type row_band: hummingbirdtech.domain.image.Band
+        :type vegetation_band: hummingbirdtech.domain.image.Band
+        :type crop_row_lines: List[shapely.geometry.MultiLineString]
```
Comment (BlayneJLees) [inline hummingbirdtech/planting_lines/gap_detection/report.py:65]: Think these should just be `LineString`

File: hummingbirdtech/planting_lines/gap_detection/report.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            for sample_window in random_windows
+        ]
+
+    def _get_sample(self, rgb_image, row_band, vegetation_band, crop_row_lines, gap_lines, sample_window):
+        """Create a single image sample.
+
+        :type rgb_image: hummingbirdtech.domain.image.Image
+        :type row_band: hummingbirdtech.domain.image.Band
+        :type vegetation_band: hummingbirdtech.domain.image.Band
+        :type crop_row_lines: List[shapely.geometry.MultiLineString]
```
Comment (BlayneJLees) [inline hummingbirdtech/planting_lines/gap_detection/report.py:85]: Same here

File: hummingbirdtech/planting_lines/gap_detection/runnable.py [hunk 3]
```diff
@@ -112,3 +130,67 @@ def run(self, survey):
             geographic_gap_lines=geographic_gap_lines,
             gap_line_lengths=gap_line_lengths,
         )
+        if report:
+            self._logger.info("Generating report for survey {survey}".format(survey=survey.identity))
+            self.create_report(
+                survey=survey,
+                rgb_image=image,
+                row_band=self._line_creator_runnable._row_segmenter.segments(self._get_inverse_cive_band(image)),
```
Comment (BlayneJLees) [inline hummingbirdtech/planting_lines/gap_detection/runnable.py:138]: `row band` come from the cive band not the inverse cive band and vegetation band is the inverse cive band.

**Comment by aj-hitchins** [comment]
Tests have passed here on re-trigger:
https://jenkins.hummingbird-tech.io/job/Testing/job/Python/job/hummingbirdtech-planting-lines/24/console

**Comment by aj-hitchins** [comment]
Tests are passing now"

File: CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased](https://github.com/HummingbirdTechGroup/hummingbirdtech-planting-lines/tree/develop)
 
 ### Added
+- Reports
```
Comment (VirginieBfd) [inline CHANGELOG.md:17]: A version has been released, therefore the CHANGELOG should reflect it.

File: hummingbirdtech/planting_lines/__init__.py [hunk 5]
```diff
@@ -5,7 +5,7 @@
 from .line_reconstruction.dic import parameters as line_reconstruction_parameters
 
 
-__version__ = "1.4.1"
+__version__ = "1.5.1"
```
Comment (VirginieBfd) [inline hummingbirdtech/planting_lines/__init__.py:8]: 1.5.0 instead isn't?

File: hummingbirdtech/planting_lines/gap_detection/report.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        :type logger: logbook.Logger
+        :type coordinate_converter:
+        :type report_survey_metadata_serialiser: hummingbirdtech.report.serialiser.SurveyMetadataSerialiser
+        """
+        self._logger = logger
+        self._coordinate_converter = coordinate_converter
+        self._report_survey_metadata_serialiser = report_survey_metadata_serialiser
+        self._window_width = 35
+        self._window_height = 35
+        self._nsamples = 10
```
Comment (VirginieBfd) [inline hummingbirdtech/planting_lines/gap_detection/report.py:21]: Do we want to keep the parameters here or in `parameters.py`?

File: hummingbirdtech/planting_lines/__init__.py (hunk 5 above)
Comment (LTMeyer) [inline hummingbirdtech/planting_lines/__init__.py:8]: Version should be `1.5.0`

File: hummingbirdtech/planting_lines/gap_detection/report.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        }
+
+    def _get_statistics(self, gap_lines):
+        """
+        :type gap_lines: List[shapely.geometry.LineString]
+        :rtype: dict
+        """
+        values = np.array([line.length for line in gap_lines])
+        return {
+            "Gap Line Length Maximum": "{max:.2f} meters".format(max=np.max(values)),
```
Comment (VirginieBfd) [inline hummingbirdtech/planting_lines/gap_detection/report.py:57]: Can this be a parameter?

File: hummingbirdtech/planting_lines/gap_detection/report.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        """
+        :type lines: List[shapely.geometry.LineString]
+        :rtype: float
+        """
+
+    def _get_gap_line_coverage(self, lines):
+        """
+        :type lines: List[shapely.geometry.LineString]
+        :rtype: float
+        """
```
Comment (VirginieBfd) [inline hummingbirdtech/planting_lines/gap_detection/report.py:75]: Are we planning on displaying the following?

Comment (LTMeyer) [inline hummingbirdtech/planting_lines/gap_detection/report.py:75]: These functions do nothing yet. Is this expected?

File: hummingbirdtech/planting_lines/gap_detection/report.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        :type gap_lines: List[shapely.geometry.LineString]
+        :type sample_window: hummingbirdtech.domain.image.Window
+        :rtype List[dict]
+        """
+        # band as type int16
+
+        sample_image = Image(
+            [
+                band.crop(window=sample_window)
+                .as_type(np.dtype("int16"))
```
Comment (LTMeyer) [inline hummingbirdtech/planting_lines/gap_detection/report.py:186]: Why do we use `int16` here? Does it not create an issue for visualisation, matplotlib might expect `float` or `uint8`.

File: hummingbirdtech/planting_lines/gap_detection/report.py (hunk 8 above)
Comment (BlayneJLees) [reply hummingbirdtech/planting_lines/gap_detection/report.py:75]: @aj-hitchins I assume these can be deleted?

File: hummingbirdtech/planting_lines/gap_detection/report.py (hunk 7 above)
Comment (aj-hitchins) [reply hummingbirdtech/planting_lines/gap_detection/report.py:57]: Are you referring to the `:.2f`?

File: hummingbirdtech/planting_lines/gap_detection/report.py (hunk 8 above)
Comment (aj-hitchins) [reply hummingbirdtech/planting_lines/gap_detection/report.py:75]: Yes please.

File: hummingbirdtech/planting_lines/gap_detection/report.py (hunk 9 above)
Comment (aj-hitchins) [reply hummingbirdtech/planting_lines/gap_detection/report.py:186]: Yeah you need to have int for matplotlib to display in rgb.

File: hummingbirdtech/planting_lines/gap_detection/report.py (hunk 6 above)
Comment (aj-hitchins) [reply hummingbirdtech/planting_lines/gap_detection/report.py:21]: I don't see it changing any time soon, if it needs to we can add it to DIC.

File: hummingbirdtech/planting_lines/gap_detection/report.py (hunk 9 above)
Comment (LTMeyer) [reply hummingbirdtech/planting_lines/gap_detection/report.py:174]: Yes but why 16 bits and not unsigned 8?
I beleive matplotlib will not handle correctly `int16`. Matplotlib will however understand either `uint8`, `float32`.

Comment (aj-hitchins) [reply hummingbirdtech/planting_lines/gap_detection/report.py:174]: It displays correctly in the report. Link here:
https://drive.google.com/file/d/1ryTxAkjvOXxez2srLiU-2mrYC_oMGre1/view?usp=sharing

**Comment by aj-hitchins** [comment]
@LTMeyer https://github.com/HummingbirdTechGroup/hummingbirdtech-planting-lines/pull/20#discussion_r420690371

**Pull request merged**